### PR TITLE
Small cmd refactoring to use the same cli helper as docker/cli

### PR DESCRIPTION
**Pull request by vdemeester**
- This makes it easier to unit test sub-commands.
- This will help us share more code with docker/cli in the near
  future.

This also contains some `refactoring` on `packager` and `render`
packages, mainly for input/output handling.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: internal/packager/init.go [hunk 1]
```diff
@@ -70,7 +71,13 @@ func Init(name string, composeFile string, description string, maintainers []str
 		return err
 	}
 	defer os.RemoveAll(temp)
-	return Merge(temp, dirName)
+	var target io.Writer
+	target, err = os.Create(dirName)
```
Comment (silvin-lubecki) [inline internal/packager/init.go]: nit: `target, err :=` is valid too

File: cmd/docker-app/split.go [hunk 2]
```diff
@@ -15,11 +15,16 @@ func splitCmd() *cobra.Command {
...
+			if err != nil {
+				return err
+			}
+			defer cleanup()
+			return packager.Split(appname, splitOutputDir)
 		},
 	}
 	if internal.Experimental == "on" {
-		cmd.Flags().StringVarP(&splitOutputDir, "output", "o", "-", "Output directory")
+		cmd.Flags().StringVarP(&splitOutputDir, "output", "o", ".", "Output directory")
```
Comment (silvin-lubecki) [inline cmd/docker-app/split.go]: You replaced `s/-/.`, is that what you wanted?

File: cmd/docker-app/main.go [hunk 3]
```diff
@@ -2,10 +2,19 @@ package main
...
+	"github.com/sirupsen/logrus"
 )
 
 func main() {
-	cmd := newRootCmd()
+	// Set terminal emulation based on platform as required.
+	stdin, stdout, stderr := term.StdStreams()
+	logrus.SetOutput(stderr)
+
+	dockerCli := command.NewDockerCli(stdin, stdout, stderr, false)
```
Comment (silvin-lubecki) [inline cmd/docker-app/main.go]: No content trust for the moment?

File: internal/renderer/inspect.go [hunk 4]
```diff
@@ -14,7 +15,7 @@ import (
 )
 
 // Inspect dumps the metadata of an app
-func Inspect(appname string) error {
+func Inspect(out io.Writer, appname string) error {
```
Comment (silvin-lubecki) [inline internal/renderer/inspect.go]: I think you wanted to use it with `w := tabwriter.NewWriter(os.Stdout`

File: cmd/docker-app/split.go (hunk 2 above)
Comment (silvin-lubecki) [inline cmd/docker-app/split.go]: Did you replace `s/-/.` intentionally?

File: cmd/docker-app/root.go [hunk 5]
```diff
@@ -5,35 +5,14 @@ import (
...
-		packCmd(),
-		pullCmd(),
-		splitCmd(),
-		unpackCmd(),
-	}
-)
-
 // rootCmd represents the base command when called without any subcommands
-func newRootCmd() *cobra.Command {
+// FIXME(vdemeester) use command.Cli interface
```
Comment (silvin-lubecki) [inline cmd/docker-app/root.go]: Is it the Initialize in the Deploy command which prevent you to use command.Cli interface?

File: internal/packager/init.go (hunk 1 above)
Comment (silvin-lubecki) [inline internal/packager/init.go]: nit: `target, err :=` is valid too 😄

File: cmd/docker-app/main.go (hunk 3 above)
Comment (vdemeester) [reply cmd/docker-app/main.go]: @silvin-lubecki yes, it wasn't supported on `docker/app` before this PR, it won't after :stuck_out_tongue:

File: cmd/docker-app/root.go (hunk 5 above)
Comment (vdemeester) [reply cmd/docker-app/root.go]: @silvin-lubecki yes :sweat_smile: I'm preparing to bump the `docker/cli` and upstream a lot of changes, that's why I put the FIXME :angel: a way to say "it's gonna go at some point :wink:"

File: cmd/docker-app/split.go (hunk 2 above)
Comment (vdemeester) [reply cmd/docker-app/split.go]: Yes, it's a folder so `-` doesn't make any sense (and it's not supported by `packager.Split` later on)

File: internal/renderer/inspect.go (hunk 4 above)
Comment (vdemeester) [reply internal/renderer/inspect.go]: oh good point, forgot to do something in there :sweat_smile:

File: cmd/docker-app/main.go (hunk 3 above)
Comment (silvin-lubecki) [reply cmd/docker-app/main.go]: Right, but do we want it? Do we need to add an issue just to remember it?

**Comment by silvin-lubecki** [comment]
Sorry for the duplication in comments, GH had some weird bug 🤔

File: cmd/docker-app/main.go (hunk 3 above)
Comment (vdemeester) [reply cmd/docker-app/main.go]: @silvin-lubecki in short-term, in long-term, invoking @garethr but I would guess yes :angel:

Comment (garethr) [reply cmd/docker-app/main.go]: Yes, we'll definitely look at the use of content trust for applications in the future.

File: cmd/docker-app/pack.go [hunk 6]
```diff
@@ -1,20 +1,43 @@
...
+				return err
+			}
+			defer cleanup()
+			var target io.Writer
+			if packOutputFile == "-" {
+				if terminal.IsTerminal(int(dockerCli.Out().FD())) {
+					return fmt.Errorf("Refusing to output to a terminal, use a shell redirect or the '-o' option")
+				}
+			} else {
+				target, err = os.Create(packOutputFile)
```
Comment (mat007) [inline cmd/docker-app/pack.go]: No defer Close here?

File: cmd/docker-app/pack.go [hunk 7]
```diff
@@ -1,20 +1,43 @@
...
+			appname := firstOrEmpty(args)
+			appname, cleanup, err := packager.Extract(appname)
+			if err != nil {
+				return err
+			}
+			defer cleanup()
+			var target io.Writer
+			if packOutputFile == "-" {
+				if terminal.IsTerminal(int(dockerCli.Out().FD())) {
+					return fmt.Errorf("Refusing to output to a terminal, use a shell redirect or the '-o' option")
```
Comment (mat007) [inline cmd/docker-app/pack.go]: Shouldn't error message be all lower case?

File: internal/packager/init.go (hunk 1 above)
Comment (mat007) [inline internal/packager/init.go]: Why the var? Can't we let `target` be of type `*os.File`?

File: cmd/docker-app/pack.go [hunk 8]
```diff
@@ -1,20 +1,43 @@
...
 		Args:  cli.RequiresMaxArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			return packager.Pack(firstOrEmpty(args), packOutputFile)
+			appname := firstOrEmpty(args)
+			appname, cleanup, err := packager.Extract(appname)
+			if err != nil {
+				return err
+			}
+			defer cleanup()
+			var target io.Writer
```
Comment (mat007) [inline cmd/docker-app/pack.go]: Would be easier to defer Close with a `var target *os.File` I believe.

File: internal/packager/split.go [hunk 9]
```diff
@@ -9,12 +9,7 @@ import (
 
 // Split converts an app package to the split version
 func Split(appname string, outputDir string) error {
-	appname, cleanup, err := Extract(appname)
-	if err != nil {
-		return err
-	}
-	defer cleanup()
-	err = os.Mkdir(outputDir, 0755)
+	err := os.Mkdir(outputDir, 0755)
```
Comment (mat007) [inline internal/packager/split.go]: Shouldn't that actually be a `os.MkdirAll` ?

**Pull request merged**